### ISSUE TITLE
How to refresh token in Laravel 5

### ISSUE THREAD

**Issue by bhavikji** [djchen/oauth2-fitbit#36]
Hi, I'm using oauth2-fitbit in my laravel application on laravel 5.3 and till now I have implemented first part of your code it worked like charm for me and I have saved my access token into db now I need to implement middleware that will keep a check over tokens whether it has expired or not. If the token is expired then simply the middleware will have some code to refresh it and save it into db and by filtering the flow of app will resume again. 

something like this,

public function handle( $request, Closure $next ) {
		$provider            = new Fitbit( [
			'clientId'     => env( 'FITBIT_KEY' ),
			'clientSecret' => env( 'FITBIT_SECRET' ),
			'redirectUri'  => env( 'FITBIT_REDIRECT_URI' )
		] );
		$patientID = Auth::user()->Patient->PatientID;
		$getPatient = FitbitMaster::where('patient_id',$patientID)->first();
		$existingAccessToken = $getPatient->fitbit_token;

		if ( $existingAccessToken->hasExpired() ) {
			$newAccessToken = $provider->getAccessToken( 'refresh_token', [
				'refresh_token' => $existingAccessToken->getRefreshToken()
			] );
			$token = $newAccessToken->getToken();
			FitbitMaster::where('patient_id',$patientID)
			            ->update(['fitbit_token',$token]);
		}

		return $next( $request );
	}

In this middleware I'm fetching token from db and then I'm checking it that whether the token is expired or not but when I'm doing so, I'm getting this error 

unable to call to a member function on string $hasExpired. Well, I'm not much familiar with your package but following your code 

$provider = new djchen\OAuth2\Client\Provider\Fitbit([
    'clientId'          => '{fitbit-oauth2-client-id}',
    'clientSecret'      => '{fitbit-client-secret}',
    'redirectUri'       => 'https://example.com/callback-url'
]);

$existingAccessToken = getAccessTokenFromYourDataStore();

if ($existingAccessToken->hasExpired()) {
    $newAccessToken = $provider->getAccessToken('refresh_token', [
        'refresh_token' => $existingAccessToken->getRefreshToken()
    ]);

    // Purge old access token and store new access token to your data store.
}

I haven't got getAccessTokenFromYourDataStore() and also this 'djchen\OAuth2\Client\Provider\Fitbit' class in my vendor. Where I'm wrong here? Where is this function getAccessTokenFromYourDataStore() what this function is doing or what it is returning? 

Also, If I am doing something wrong then what is the proper way to handling refresh token mechanism?

Thanks 
Oscar! :)

**Comment by djchen** [comment]
Hi Oscar,

The `getAccessTokenFromYourDataStore()` is something that you'll have to implement to retrieve the refresh token object from your database.

It sounds like you already have something like it with `$existingAccessToken = $getPatient->fitbit_token;`. Are you storing the `League\OAuth2\Client\Token\AccessToken` object in your database or is that just the access token string? If its the former, you should be able to call getRefreshToken() on the retrieved object to get the refresh token string to pass into
```
$provider->getAccessToken('refresh_token', [
'refresh_token' => REFRESH_TOKEN_HERE
]);
```

Otherwise, you can store the refresh token from the `League\OAuth2\Client\Token\AccessToken` object separately in your database as separate column.

**Comment by bhavikji** [comment]
Hi djchen,

I'm just storing token string which I'm getting from getToken() function. so should I call AccessToken()? and if yes then how could I call League\OAuth2\Client\Token\AccessToken object so that I could able to call refreshToken()?

**Comment by djchen** [comment]
What you want to do is store both the access token from the `getToken()` function and the refresh token from the `getRefreshToken()` method.

From the example code, its this step:
```
        // Try to get an access token using the authorization code grant.
        $accessToken = $provider->getAccessToken('authorization_code', [
            'code' => $_GET['code']
        ]);

        // We have an access token, which we may use in authenticated
        // requests against the service provider's API.
        echo $accessToken->getToken() . "\n";
        echo $accessToken->getRefreshToken() . "\n";
```

When you get the $accessToken (AccessToken object), you need to store both the token and the refresh token. (The 2 things that are being echo'ed). You'll need to do the same for both when you refresh the token.

**Comment by bhavikji** [comment]
Ok I did this I have stored both refresh token and token now I'm confused here how should I call hasExpired()  since I'm getting string from getRefreshToken() and that's what I'm storing also what if I try to use this thing with cron as per requirement I need to call whole mechanism using artisan command.

**Comment by djchen** [comment]
In order to do that, you need to serialize and store the AccessToken object into your database and deserialize it when you read it back out.

**Comment by bhavikji** [comment]
yeah finally I fixed this thanks for enlighten me. cheers! :)

**Comment by djchen** [comment]
Awesome!

**Comment by bhavikji** [comment]
Hi, I got something new here, like you said yesterday about refresh token mechanism I did exactly the same and it worked perfectly for me.

But Now I'm getting a new error and for that I'm illustrating the process of my app

1. I'm authorizing the person and saving his AccessToken object and also registering into my app.
2. After registration I'm assuming that he has AccessToken and now he doesn't to connect from fitbit again.
2. Then, I'm checking whether is access token is expired or not.
3. If it is expired then hasExpired() will be called and will try to get a new token
4. All these steps I'm doing from Laravel Console as I will fire a command and then the command will take care of everything.

Now the problem is I'm getting something like invalid_refresh_token error 
Under what circumstances does this error occurs?

Is it possible that after one time connection I can retrieve/refresh the token from the Access Token object from my data store??
